[DLPack] add NumPy exchange tests.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150691
* #150218
* #150217
* __->__ #150216

This PR resolves an old TODO that requested NumPy DLPack exchange tests
once version 1.22 was required.

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

msaroufim reviewed (approved):
could be a seperate PR but I'd really love to also see some tests testing moving gpu tensors to a framework that also supports gpus

albanD commented on test/test_dlpack.py:
```diff
@@ -305,6 +306,42 @@ def test(device, **kwargs):
+        # Transform the NumPy array back using DLPack.
+        res = from_dlpack(t_arr)
+
+        self.assertEqual(t, res.cpu())
```
The .cpu() here is a noop I guess?

albanD commented on test/test_dlpack.py:
```diff
@@ -305,6 +306,42 @@ def test(device, **kwargs):
+        res = from_dlpack(t_arr)
+
+        self.assertEqual(t, res.cpu())
+        # If device is CPU (same as our original tensor), then they
```
outdated comment?

albanD reviewed (approved):
Small updates needed, but SGTM otherwise

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] add NumPy exchange tests."


This PR resolves an old TODO that requested NumPy DLPack exchange tests
once version 1.22 was required.

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "[DLPack] add NumPy exchange tests."


This PR resolves an old TODO that requested NumPy DLPack exchange tests
once version 1.22 was required.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] add NumPy exchange tests."


This PR resolves an old TODO that requested NumPy DLPack exchange tests
once version 1.22 was required.

[ghstack-poisoned]
